scipy/optimize/_hessian_update_strategy.py

```diff
@@ -188,15 +188,51 @@ def update(self, delta_x, delta_grad):
             return
         if self.first_iteration:
             # Get user specific scale
-            if self.init_scale == "auto":
+            if isinstance(self.init_scale, str) and self.init_scale == "auto":
                 scale = self._auto_scale(delta_x, delta_grad)
             else:
-                scale = float(self.init_scale)
-            # Scale initial matrix with ``scale * np.eye(n)``
+                scale = self.init_scale
+
+            # Check for complex: numpy will silently cast a complex array to
+            # a real one but not so for scalar as it raises a TypeError.
+            # Checking here brings a consistent behavior.
+            replace = False
+            if np.size(scale) == 1:
+                # to account for the legacy behavior having the exact same cast
+                scale = float(scale)
+            elif np.iscomplexobj(scale):
+                raise TypeError("init_scale contains complex elements, "
+                                "must be real.")
+            else:  # test explicitly for allowed shapes and values
+                replace = True
+                scale = np.array(scale)  # copy, will replace the original
+                if self.approx_type == 'hess':
+                    shape = np.shape(self.B)
+                else:
+                    shape = np.shape(self.H)
+
+                # it has to match the shape of the matrix for the multiplication,
+                # no implicit broadcasting is allowed
+                if shape != np.shape(scale):
+                    raise ValueError("If init_scale is an array, it must have the "
+                                     f"dimensions of the matrix: {shape}.")
```

Comment:
I've extended the error message a bit, but not sure what exactly could be told?

(I've used matrix as in "hessian matrix", but replaced with "hess/inv_hess" as this is also used as the type keyword)

Reply:
My bad. What I meant was to actually tell the user what exactly is wrong and what has to be the right thing. So if we already detected the array with the wrong shape we can say so

```python
                if shape != np.shape(scale):
                    raise ValueError("If init_scale is an array, it must have the "
                                     f"dimensions of the matrix ({n}, {n}) but got {shape}.")

```

if it fits into PEP8 limit otherwise with a line break somewhere

Reply:
Hm, but that's exactly what it's doing right now I think? `shape` _is_ `({n}, {n})`. I've added that it also prints out the user provided array.

It's also tested in plait string in the tests

Thanks for reviewing and the scrutiny btw, appreciate it!

Reply:
Yeah I missed your change while I was typing my diatribe. Thank you for the rapid changes. Much appreciated. I'll mark it as resolved.